### PR TITLE
feat: introducing warehouse repo withTx

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

- Introducing `WithTx` to use wherever transactions are used.
- Also, this is a follow-up on #4045 where I am doing a cleanup for updating fields for the uploads table. `WithTx` is required during reporting where we update fields and report to reporting in the same transaction.

## Linear Ticket

- Resolves PIPE-462

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/internal/repo/repo.go [hunk 1]
```diff
@@ -18,3 +19,21 @@ func WithNow(now func() time.Time) Opt {
...
+
+func (r *repo) WithTx(f func(tx *sqlmiddleware.Tx) error) error {
+	tx, err := r.db.Begin()
+	if err != nil {
+		return fmt.Errorf("begin transaction: %w", err)
+	}
+
+	if err := f(tx); err != nil {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			return fmt.Errorf("rollback transaction for %w: %w", err, rollbackErr)
```
Comment (fracasula) [inline warehouse/internal/repo/repo.go:31]: This is like Nando's double :chicken: wrap but with errors :sweat_smile:

**Review by fracasula — APPROVED** [review]

File: warehouse/internal/repo/repo.go (hunk 1 above)
Comment (achettyiitr) [reply warehouse/internal/repo/repo.go:31]: I am a bit skeptical here. Should we ignore rollback error?

Comment (fracasula) [reply warehouse/internal/repo/repo.go:31]: Nah, it's OK. It works perfectly fine and it is possible to unwrap both errors :+1:

File: warehouse/internal/repo/load.go [hunk 2]
```diff
@@ -60,58 +59,47 @@ func (repo *LoadFiles) DeleteByStagingFiles(ctx context.Context, stagingFileIDs
...
-	}
-	err = txn.Commit()
-	if err != nil {
-		return fmt.Errorf(`inserting load files: commit: %w`, err)
-	}
-	return
+		defer func() { _ = stmt.Close() }()
+
+		for _, loadFile := range loadFiles {
+			metadata := fmt.Sprintf(`{"content_length": %d, "destination_revision_id": %q, "use_rudder_storage": %t}`, loadFile.ContentLength, loadFile.DestinationRevisionID, loadFile.UseRudderStorage)
```
Comment (lvrach) [inline warehouse/internal/repo/load.go:85]: [minor] I would prefer a proper` json.Marshal()`, IMO is cleaner and safer

File: warehouse/internal/repo/repo.go (hunk 1 above)
Comment (lvrach) [reply warehouse/internal/repo/repo.go:33]: I wasn't aware that is possible to wrap two errors that the same time, cool

**Review by lvrach — APPROVED** [review]

**Pull request merged**